src/main/java/fi/otavanopisto/kuntaapi/server/integrations/AbstractIdProvider.java

```diff
@@ -168,35 +170,59 @@ public AnnouncementId translate(AnnouncementId announcementId, String target) {
+    return getSource().equals(baseId.getSource()) && KuntaApiConsts.IDENTIFIER_NAME.equals(target);
   }
   
-  private boolean toSourceId(Id id, String target) {
-    return target.equals(getSource()) && KuntaApiConsts.IDENTIFIER_NAME.equals(id.getSource());
+  private boolean toSourceId(BaseId baseId, String target) {
+    return target.equals(getSource()) && KuntaApiConsts.IDENTIFIER_NAME.equals(baseId.getSource());
   }
 
-  private <T extends Id> T translateToSourceId(T id, IdType type, Class<T> idClass) {
+  private <T extends BaseId> T translateToSourceId(T id, IdType type, Class<T> idClass) {
     Identifier identifier;
     identifier = identifierController.findIdentifierByTypeSourceAndKuntaApiId(type, getSource(), id.getId());
     if (identifier != null) {
-      return createId(idClass, getSource(), identifier.getSourceId());
+      if (isOrganizationBaseId(idClass)) {
+        return createOrganizationBaseId(idClass, getSource(), identifier.getSourceId(), identifier.getOrganizationKuntaApiId());
+      } else {
+        return createBaseId(idClass, getSource(), identifier.getSourceId());
+      }
     }
     
     return null;
   }
 
-  private <T extends Id> T translateToKuntaApiId(T id, Class<T> idClass) {
+  private <T extends BaseId> T translateToKuntaApiId(T id, Class<T> idClass) {
     Identifier identifier;
     identifier = identifierController.findIdentifierById(id);
     if (identifier != null) {
-      return createId(idClass, KuntaApiConsts.IDENTIFIER_NAME, identifier.getKuntaApiId());
+      if (isOrganizationBaseId(idClass)) {
+        return createOrganizationBaseId(idClass, KuntaApiConsts.IDENTIFIER_NAME, identifier.getKuntaApiId(), identifier.getOrganizationKuntaApiId()); 
+      } else {
+        return createBaseId(idClass, KuntaApiConsts.IDENTIFIER_NAME, identifier.getKuntaApiId());
+      }
     }
     
     return null;
   }
   
-  private <T extends Id> T createId(Class<T> idClass, String source, String id) {
+  private boolean isOrganizationBaseId(Class<? extends BaseId> idClass) {
+    return OrganizationBaseId.class.isAssignableFrom(idClass);
+  }
+  private <T extends BaseId> T createOrganizationBaseId(Class<T> idClass, String source, String id, String organizationKuntaApiId) {
+    Constructor<T> idConstructor;
+    try {
+      OrganizationId organizationId = new OrganizationId(KuntaApiConsts.IDENTIFIER_NAME, organizationKuntaApiId);
+      idConstructor = idClass.getDeclaredConstructor(OrganizationId.class, String.class, String.class);
+      return idConstructor.newInstance(organizationId, source, id);
+    } catch (NoSuchMethodException | SecurityException | InstantiationException | IllegalAccessException | IllegalArgumentException | InvocationTargetException e) {
+      logger.log(Level.SEVERE, "Failed to construct id", e);
+    }
+    
+    return null;
+  }
+  
+  private <T extends BaseId> T createBaseId(Class<T> idClass, String source, String id) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this unused private "createBaseId" method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AUnusedPrivateMethod)